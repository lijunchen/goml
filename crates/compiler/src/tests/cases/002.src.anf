fn main() -> Unit {
  let a/0 = (true, false) in
  let x0 = a/0.0 in
  let x1 = a/0.1 in
  let b/1 = match x1 {
    true => {
      match x0 {
        true => {
          (false, false)
        },
        false => {
          (true, false)
        },
      }
    },
    false => {
      match x0 {
        true => {
          (false, true)
        },
        false => {
          (true, true)
        },
      }
    },
  } in
  let x2 = b/1.0 in
  let x3 = b/1.1 in
  let w/2 = x3 in
  let b_1/3 = w/2 in
  let mtmp4 = (true, b_1/3) in
  let x5 = mtmp4.0 in
  let x6 = mtmp4.1 in
  let c/4 = match x6 {
    true => {
      match x5 {
        true => {
          let t7 = int_to_string(3) in
          print(t7)
        },
        false => {
          let t8 = int_to_string(1) in
          print(t8)
        },
      }
    },
    false => {
      match x5 {
        true => {
          let t9 = int_to_string(2) in
          print(t9)
        },
        false => {
          let t10 = int_to_string(0) in
          print(t10)
        },
      }
    },
  } in
  let t11 = unit_to_string(c/4) in
  print(t11)
}