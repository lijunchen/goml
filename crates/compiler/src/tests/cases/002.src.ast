fn main() {
    let a = (true, false) in
    let b = match a {
        (false, false) => (true, true),
        (false, true) => (true, false),
        (true, false) => (false, true),
        (true, true) => (false, false),
    } in
    let c = match (true, b.1) {
        (false, false) => print(int_to_string(0)),
        (false, true) => print(int_to_string(1)),
        (true, false) => print(int_to_string(2)),
        (true, true) => print(int_to_string(3)),
    } in
    print(unit_to_string(c))
}

