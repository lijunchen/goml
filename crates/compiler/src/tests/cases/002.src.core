fn main() -> Unit {
  let a/0 = (true, false) in
  let b/1 = let x0 = a/0.0 in
  let x1 = a/0.1 in
  match x1 {
    true => {
      match x0 {
        true => {
          (false, false)
        },
        false => {
          (true, false)
        },
      }
    },
    false => {
      match x0 {
        true => {
          (false, true)
        },
        false => {
          (true, true)
        },
      }
    },
  } in
  let c/2 = let mtmp2 = (true, b/1.1) in
  let x3 = mtmp2.0 in
  let x4 = mtmp2.1 in
  match x4 {
    true => {
      match x3 {
        true => {
          print(int_to_string(3))
        },
        false => {
          print(int_to_string(1))
        },
      }
    },
    false => {
      match x3 {
        true => {
          print(int_to_string(2))
        },
        false => {
          print(int_to_string(0))
        },
      }
    },
  } in
  print(unit_to_string(c/2))
}