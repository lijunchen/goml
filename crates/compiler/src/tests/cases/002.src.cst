FILE@0..508
  FN@0..508
    FnKeyword@0..2 "fn"
    Whitespace@2..3 " "
    Lident@3..7 "main"
    PARAM_LIST@7..10
      LParen@7..8 "("
      RParen@8..9 ")"
      Whitespace@9..10 " "
    BLOCK@10..508
      LBrace@10..11 "{"
      Whitespace@11..16 "\n    "
      EXPR_LET@16..506
        LetKeyword@16..19 "let"
        Whitespace@19..20 " "
        PATTERN_VARIABLE@20..22
          Lident@20..21 "a"
          Whitespace@21..22 " "
        Eq@22..23 "="
        Whitespace@23..24 " "
        EXPR_LET_VALUE@24..38
          EXPR_TUPLE@24..38
            LParen@24..25 "("
            EXPR_BOOL@25..29
              TrueKeyword@25..29 "true"
            Comma@29..30 ","
            Whitespace@30..31 " "
            EXPR_BOOL@31..36
              FalseKeyword@31..36 "false"
            RParen@36..37 ")"
            Whitespace@37..38 " "
        InKeyword@38..40 "in"
        Whitespace@40..45 "\n    "
        EXPR_LET_BODY@45..506
          EXPR_LET@45..506
            LetKeyword@45..48 "let"
            Whitespace@48..49 " "
            PATTERN_VARIABLE@49..51
              Lident@49..50 "b"
              Whitespace@50..51 " "
            Eq@51..52 "="
            Whitespace@52..53 " "
            EXPR_LET_VALUE@53..229
              EXPR_MATCH@53..229
                MatchKeyword@53..58 "match"
                Whitespace@58..59 " "
                EXPR_LIDENT@59..61
                  Lident@59..60 "a"
                  Whitespace@60..61 " "
                MATCH_ARM_LIST@61..229
                  LBrace@61..62 "{"
                  Whitespace@62..71 "\n        "
                  MATCH_ARM@71..101
                    PATTERN_TUPLE@71..86
                      LParen@71..72 "("
                      PATTERN_BOOL@72..77
                        FalseKeyword@72..77 "false"
                      Comma@77..78 ","
                      Whitespace@78..79 " "
                      PATTERN_BOOL@79..84
                        FalseKeyword@79..84 "false"
                      RParen@84..85 ")"
                      Whitespace@85..86 " "
                    FatArrow@86..88 "=>"
                    Whitespace@88..89 " "
                    EXPR_TUPLE@89..101
                      LParen@89..90 "("
                      EXPR_BOOL@90..94
                        TrueKeyword@90..94 "true"
                      Comma@94..95 ","
                      Whitespace@95..96 " "
                      EXPR_BOOL@96..100
                        TrueKeyword@96..100 "true"
                      RParen@100..101 ")"
                  Comma@101..102 ","
                  Whitespace@102..111 "\n        "
                  MATCH_ARM@111..141
                    PATTERN_TUPLE@111..125
                      LParen@111..112 "("
                      PATTERN_BOOL@112..117
                        FalseKeyword@112..117 "false"
                      Comma@117..118 ","
                      Whitespace@118..119 " "
                      PATTERN_BOOL@119..123
                        TrueKeyword@119..123 "true"
                      RParen@123..124 ")"
                      Whitespace@124..125 " "
                    FatArrow@125..127 "=>"
                    Whitespace@127..128 " "
                    EXPR_TUPLE@128..141
                      LParen@128..129 "("
                      EXPR_BOOL@129..133
                        TrueKeyword@129..133 "true"
                      Comma@133..134 ","
                      Whitespace@134..135 " "
                      EXPR_BOOL@135..140
                        FalseKeyword@135..140 "false"
                      RParen@140..141 ")"
                  Comma@141..142 ","
                  Whitespace@142..151 "\n        "
                  MATCH_ARM@151..181
                    PATTERN_TUPLE@151..165
                      LParen@151..152 "("
                      PATTERN_BOOL@152..156
                        TrueKeyword@152..156 "true"
                      Comma@156..157 ","
                      Whitespace@157..158 " "
                      PATTERN_BOOL@158..163
                        FalseKeyword@158..163 "false"
                      RParen@163..164 ")"
                      Whitespace@164..165 " "
                    FatArrow@165..167 "=>"
                    Whitespace@167..168 " "
                    EXPR_TUPLE@168..181
                      LParen@168..169 "("
                      EXPR_BOOL@169..174
                        FalseKeyword@169..174 "false"
                      Comma@174..175 ","
                      Whitespace@175..176 " "
                      EXPR_BOOL@176..180
                        TrueKeyword@176..180 "true"
                      RParen@180..181 ")"
                  Comma@181..182 ","
                  Whitespace@182..191 "\n        "
                  MATCH_ARM@191..221
                    PATTERN_TUPLE@191..204
                      LParen@191..192 "("
                      PATTERN_BOOL@192..196
                        TrueKeyword@192..196 "true"
                      Comma@196..197 ","
                      Whitespace@197..198 " "
                      PATTERN_BOOL@198..202
                        TrueKeyword@198..202 "true"
                      RParen@202..203 ")"
                      Whitespace@203..204 " "
                    FatArrow@204..206 "=>"
                    Whitespace@206..207 " "
                    EXPR_TUPLE@207..221
                      LParen@207..208 "("
                      EXPR_BOOL@208..213
                        FalseKeyword@208..213 "false"
                      Comma@213..214 ","
                      Whitespace@214..215 " "
                      EXPR_BOOL@215..220
                        FalseKeyword@215..220 "false"
                      RParen@220..221 ")"
                  Comma@221..222 ","
                  Whitespace@222..227 "\n    "
                  RBrace@227..228 "}"
                  Whitespace@228..229 " "
            InKeyword@229..231 "in"
            Whitespace@231..236 "\n    "
            EXPR_LET_BODY@236..506
              EXPR_LET@236..506
                LetKeyword@236..239 "let"
                Whitespace@239..240 " "
                PATTERN_VARIABLE@240..242
                  Lident@240..241 "c"
                  Whitespace@241..242 " "
                Eq@242..243 "="
                Whitespace@243..244 " "
                EXPR_LET_VALUE@244..474
                  EXPR_MATCH@244..474
                    MatchKeyword@244..249 "match"
                    Whitespace@249..250 " "
                    EXPR_TUPLE@250..262
                      LParen@250..251 "("
                      EXPR_BOOL@251..255
                        TrueKeyword@251..255 "true"
                      Comma@255..256 ","
                      Whitespace@256..257 " "
                      EXPR_BINARY@257..260
                        EXPR_LIDENT@257..258
                          Lident@257..258 "b"
                        Dot@258..259 "."
                        EXPR_INT@259..260
                          Int@259..260 "1"
                      RParen@260..261 ")"
                      Whitespace@261..262 " "
                    MATCH_ARM_LIST@262..474
                      LBrace@262..263 "{"
                      Whitespace@263..272 "\n        "
                      MATCH_ARM@272..313
                        PATTERN_TUPLE@272..287
                          LParen@272..273 "("
                          PATTERN_BOOL@273..278
                            FalseKeyword@273..278 "false"
                          Comma@278..279 ","
                          Whitespace@279..280 " "
                          PATTERN_BOOL@280..285
                            FalseKeyword@280..285 "false"
                          RParen@285..286 ")"
                          Whitespace@286..287 " "
                        FatArrow@287..289 "=>"
                        Whitespace@289..290 " "
                        EXPR_CALL@290..313
                          EXPR_LIDENT@290..295
                            Lident@290..295 "print"
                          ARG_LIST@295..313
                            LParen@295..296 "("
                            ARG@296..312
                              EXPR_CALL@296..312
                                EXPR_LIDENT@296..309
                                  Lident@296..309 "int_to_string"
                                ARG_LIST@309..312
                                  LParen@309..310 "("
                                  ARG@310..311
                                    EXPR_INT@310..311
                                      Int@310..311 "0"
                                  RParen@311..312 ")"
                            RParen@312..313 ")"
                      Comma@313..314 ","
                      Whitespace@314..323 "\n        "
                      MATCH_ARM@323..363
                        PATTERN_TUPLE@323..337
                          LParen@323..324 "("
                          PATTERN_BOOL@324..329
                            FalseKeyword@324..329 "false"
                          Comma@329..330 ","
                          Whitespace@330..331 " "
                          PATTERN_BOOL@331..335
                            TrueKeyword@331..335 "true"
                          RParen@335..336 ")"
                          Whitespace@336..337 " "
                        FatArrow@337..339 "=>"
                        Whitespace@339..340 " "
                        EXPR_CALL@340..363
                          EXPR_LIDENT@340..345
                            Lident@340..345 "print"
                          ARG_LIST@345..363
                            LParen@345..346 "("
                            ARG@346..362
                              EXPR_CALL@346..362
                                EXPR_LIDENT@346..359
                                  Lident@346..359 "int_to_string"
                                ARG_LIST@359..362
                                  LParen@359..360 "("
                                  ARG@360..361
                                    EXPR_INT@360..361
                                      Int@360..361 "1"
                                  RParen@361..362 ")"
                            RParen@362..363 ")"
                      Comma@363..364 ","
                      Whitespace@364..373 "\n        "
                      MATCH_ARM@373..413
                        PATTERN_TUPLE@373..387
                          LParen@373..374 "("
                          PATTERN_BOOL@374..378
                            TrueKeyword@374..378 "true"
                          Comma@378..379 ","
                          Whitespace@379..380 " "
                          PATTERN_BOOL@380..385
                            FalseKeyword@380..385 "false"
                          RParen@385..386 ")"
                          Whitespace@386..387 " "
                        FatArrow@387..389 "=>"
                        Whitespace@389..390 " "
                        EXPR_CALL@390..413
                          EXPR_LIDENT@390..395
                            Lident@390..395 "print"
                          ARG_LIST@395..413
                            LParen@395..396 "("
                            ARG@396..412
                              EXPR_CALL@396..412
                                EXPR_LIDENT@396..409
                                  Lident@396..409 "int_to_string"
                                ARG_LIST@409..412
                                  LParen@409..410 "("
                                  ARG@410..411
                                    EXPR_INT@410..411
                                      Int@410..411 "2"
                                  RParen@411..412 ")"
                            RParen@412..413 ")"
                      Comma@413..414 ","
                      Whitespace@414..423 "\n        "
                      MATCH_ARM@423..462
                        PATTERN_TUPLE@423..436
                          LParen@423..424 "("
                          PATTERN_BOOL@424..428
                            TrueKeyword@424..428 "true"
                          Comma@428..429 ","
                          Whitespace@429..430 " "
                          PATTERN_BOOL@430..434
                            TrueKeyword@430..434 "true"
                          RParen@434..435 ")"
                          Whitespace@435..436 " "
                        FatArrow@436..438 "=>"
                        Whitespace@438..439 " "
                        EXPR_CALL@439..462
                          EXPR_LIDENT@439..444
                            Lident@439..444 "print"
                          ARG_LIST@444..462
                            LParen@444..445 "("
                            ARG@445..461
                              EXPR_CALL@445..461
                                EXPR_LIDENT@445..458
                                  Lident@445..458 "int_to_string"
                                ARG_LIST@458..461
                                  LParen@458..459 "("
                                  ARG@459..460
                                    EXPR_INT@459..460
                                      Int@459..460 "3"
                                  RParen@460..461 ")"
                            RParen@461..462 ")"
                      Comma@462..463 ","
                      Whitespace@463..468 "\n    "
                      RBrace@468..469 "}"
                      Whitespace@469..474 "\n    "
                InKeyword@474..476 "in"
                Whitespace@476..481 "\n    "
                EXPR_LET_BODY@481..506
                  EXPR_CALL@481..506
                    EXPR_LIDENT@481..486
                      Lident@481..486 "print"
                    ARG_LIST@486..506
                      LParen@486..487 "("
                      ARG@487..504
                        EXPR_CALL@487..504
                          EXPR_LIDENT@487..501
                            Lident@487..501 "unit_to_string"
                          ARG_LIST@501..504
                            LParen@501..502 "("
                            ARG@502..503
                              EXPR_LIDENT@502..503
                                Lident@502..503 "c"
                            RParen@503..504 ")"
                      RParen@504..505 ")"
                      Whitespace@505..506 "\n"
      RBrace@506..507 "}"
      Whitespace@507..508 "\n"