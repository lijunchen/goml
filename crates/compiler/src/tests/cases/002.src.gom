package main

import (
    "fmt"
)

func unit_to_string(x struct{}) string {
    return "()"
}

func bool_to_string(x bool) string {
    if x {
        return "true"
    } else {
        return "false"
    }
}

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func int_neg(x int) int {
    return -x
}

func int_add(x int, y int) int {
    return x + y
}

func int_sub(x int, y int) int {
    return x - y
}

func int_mul(x int, y int) int {
    return x * y
}

func int_div(x int, y int) int {
    return x / y
}

func int_less(x int, y int) bool {
    return x < y
}

func string_add(a string, b string) string {
    return a + b
}

func string_print(s string) struct{} {
    fmt.Print(s)
    return struct{}{}
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

func missing(s string) struct{} {
    println("missing: " + s)
    panic("")
    return struct{}{}
}

type Tuple2_bool_bool struct {
    _0 bool
    _1 bool
}

func main0() struct{} {
    var ret12 struct{}
    var a__0 Tuple2_bool_bool = Tuple2_bool_bool{
        _0: true,
        _1: false,
    }
    var x0 bool = a__0._0
    var x1 bool = a__0._1
    var b__1 Tuple2_bool_bool
    switch x1 {
    case true:
        switch x0 {
        case true:
            b__1 = Tuple2_bool_bool{
                _0: false,
                _1: false,
            }
        case false:
            b__1 = Tuple2_bool_bool{
                _0: true,
                _1: false,
            }
        }
    case false:
        switch x0 {
        case true:
            b__1 = Tuple2_bool_bool{
                _0: false,
                _1: true,
            }
        case false:
            b__1 = Tuple2_bool_bool{
                _0: true,
                _1: true,
            }
        }
    }
    var x3 bool = b__1._1
    var w__2 bool = x3
    var b_1__3 bool = w__2
    var mtmp4 Tuple2_bool_bool = Tuple2_bool_bool{
        _0: true,
        _1: b_1__3,
    }
    var x5 bool = mtmp4._0
    var x6 bool = mtmp4._1
    var c__4 struct{}
    switch x6 {
    case true:
        switch x5 {
        case true:
            var t7 string = int_to_string(3)
            c__4 = string_print(t7)
        case false:
            var t8 string = int_to_string(1)
            c__4 = string_print(t8)
        }
    case false:
        switch x5 {
        case true:
            var t9 string = int_to_string(2)
            c__4 = string_print(t9)
        case false:
            var t10 string = int_to_string(0)
            c__4 = string_print(t10)
        }
    }
    var t11 string = unit_to_string(c__4)
    ret12 = string_print(t11)
    return ret12
}

func main() {
    main0()
}
