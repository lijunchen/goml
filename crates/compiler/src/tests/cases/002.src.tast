fn main() -> Unit {
  let a/0: (Bool, Bool) = (true, false) in
  let b/1: (Bool, Bool) = match (a/0 : (Bool, Bool)) {
      (false, false) => (true, true),
      (false, true) => (true, false),
      (true, false) => (false, true),
      (true, true) => (false, false),
  } in
  let b_1/3: Bool = match (b/1 : (Bool, Bool)) {
      (_ : Bool, w/2: Bool) => (w/2 : Bool),
  } in
  let c/4: Unit = match (true, (b_1/3 : Bool)) {
      (false, false) => string_print(int_to_string(0)),
      (false, true) => string_print(int_to_string(1)),
      (true, false) => string_print(int_to_string(2)),
      (true, true) => string_print(int_to_string(3)),
  } in
  string_print(unit_to_string((c/4 : Unit)))
}