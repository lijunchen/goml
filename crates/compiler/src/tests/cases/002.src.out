()
====
3()
====
# command-line-arguments
./main.go:67:13: undefined: b__1
./main.go:72:13: undefined: b__1
./main.go:80:13: undefined: b__1
./main.go:85:13: undefined: b__1
./main.go:91:19: undefined: b__1
./main.go:105:13: undefined: c__4
./main.go:108:13: undefined: c__4
./main.go:114:13: undefined: c__4
./main.go:117:13: undefined: c__4
./main.go:120:37: undefined: c__4
./main.go:120:37: too many errors
